### ISSUE TITLE
[Feature] AssumeRoleWithWebIdentity / OIDC support for accessing S3 buckets

### ISSUE THREAD

**Issue by frittentheke** [allegroai/clearml#860]
## Proposal Summary
Currently only "regular" EC2 credentials, as in `key` and `secret`, can be supplied https://clear.ml/docs/latest/docs/integrations/storage/#configuring-aws-s3

I was wondering if your code and your config does allow for OIDC credentials and other settings required to be applied? Apparently boto3 is ready for it:
* https://boto3.amazonaws.com/v1/documentation/api/latest/guide/credentials.html#assume-role-with-web-identity-provider

## Motivation
Enterprises tend to use their own, existing authentication system to also manage access to AWS resources (or Ceph RADOSGW or MinIO if they run that themselves).

* MinIO - https://min.io/docs/minio/linux/operations/external-iam/configure-openid-external-identity-management.html
* Ceph RADOSGW - https://docs.ceph.com/en/quincy/radosgw/STS

AWS' capabilities also extend into using those external identities within roles and access policies, see https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_providers_oidc_user-id.html, simiar things can be done with straight bucket policies.

Having support for OIDC / WebIdentities would allow for ClearML to be used in such a context.

**Comment by erezalg** [comment]
Hi @frittentheke,

Thanks for the suggestion!
AFAIK, this is not supported for S3 access. 

Can you elaborate what you had in mind? Adding in clearml.conf a pointer to the token file and role arn?

**Comment by frittentheke** [comment]
> Can you elaborate what you had in mind? Adding in clearml.conf a pointer to the token file and role arn?

Yes, that was kind of my idea. Just to be able to set the required options and they have the S3 client in ClearML be configured accordingly. Maybe some logging statements here and there to report the thrown errors to allow folks to debug this kind of auth properly.

**Comment by frittentheke** [comment]
@erezalg Do you see any chance this feature could be implemented to support environments using OAuth/OIDC mechanisms?

**Comment by erezalg** [comment]
@frittentheke,

Thanks! That makes sense!
So basically we'll introduce a new parameter called CLEARML_USE_AWS_TOKEN_FILE, once set, it'll signal the boto3 library to look for the corresponding AWS env vars for credentials instead of explicit credentials.

That should actually be the easiest path forward, what do you think? Would that work?

**Comment by frittentheke** [comment]
@erezalg  thanks again for picking up on this.

Actually the AWS SDK has quite a few automatisms to configure itself either by ENV vars or files in e.g. ~/.aws.

So it's likely just about allowing the SDK do work it's magic and to NOT explicitly apply the configured access and secret key. That likely disables the automatic mechanisms.

But especially in Kubernetes having controllers do the work to get tokens and mount them as volumes or inject as env vars is quite common.

**Comment by erezalg** [comment]
Thanks @frittentheke, I think this makes sense. We'll update once it's implemented or if we have further questions!

**Comment by sergeyshevch** [comment]
AWS S3 fully supports IAM Role web identities already. Also as minio go sdk

**Comment by frittentheke** [comment]
> AWS S3 fully supports IAM Role web identities already. Also as minio go sdk

Yes. As I was trying to say in https://github.com/allegroai/clearml/issues/860#issuecomment-1407600642, the capabilities are all there. It's just about not forcing the SDK to a certain auth type, but to allow for the right env vars to trigger a WebIdentity auth.

**Comment by frittentheke** [comment]
@erezalg may i ask again about implementing this?

**Comment by ainoam** [comment]
@frittentheke Have you tried setting [`use_credentials_chain`](https://clear.ml/docs/latest/docs/configs/clearml_conf/#sdkawss3) in clearml.conf?
